library/std/src/thread/mod.rs

```diff
@@ -1304,46 +1350,53 @@ impl Inner {
 /// docs of [`Builder`] and [`spawn`] for more details.
 ///
 /// [`thread::current`]: current
-pub struct Thread {
-    inner: Pin<Arc<Inner>>,
-}
+pub struct Thread(Inner);
 
 impl Thread {
     /// Used only internally to construct a thread object without spawning.
     ///
     /// # Safety
     /// `name` must be valid UTF-8.
     pub(crate) unsafe fn new(name: CString) -> Thread {
-        unsafe { Self::new_inner(ThreadName::Other(name)) }
+        unsafe { Self::new_inner(Some(name)) }
     }
 
     pub(crate) fn new_unnamed() -> Thread {
-        unsafe { Self::new_inner(ThreadName::Unnamed) }
+        unsafe { Self::new_inner(None) }
     }
 
-    // Used in runtime to construct main thread
-    pub(crate) fn new_main() -> Thread {
-        unsafe { Self::new_inner(ThreadName::Main) }
+    /// Used in runtime to construct main thread
+    ///
+    /// # Safety
+    ///
+    /// This must only ever be called once, and must be called on the main thread.
+    pub(crate) unsafe fn new_main() -> Thread {
+        // Safety: Caller responsible for holding the mutable invariant and
```

Comment:
What does "holding the mutable invariant" mean? Do you mean "upholding" the invariant? What's the "mutable" invariant?

Reply:
I did mean upholding, the "mutable" invariant I was referring to is the "no existing references when creating a &mut" rule. I have written that really poorly, should this be fixed in a followup or do you want to delay this merging?

Reply:
Let's use the opportunity to do this pre-merging.